### PR TITLE
Fix: backport patch for CVE-2021-33391 for tidy

### PR DESCRIPTION
**Pull request by SeanDougherty**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

This adds a bespoke backport patch to fix the Critical [CVE-2021-33391](https://nvd.nist.gov/vuln/detail/CVE-2021-33391). This patch was created because the two PRs that originally introduce this fix to the upstream, do not cleanly apply to the currently latest stable version of the upstream (v5.8.0). The two PRs mentioned are ([PR-984](https://github.com/htacg/tidy-html5/pull/984), [PR-981](https://github.com/htacg/tidy-html5/pull/981/files)). PR-984 fixes the vulnerability, while parts of PR-981 (namely the changes to lexer.h and lexer.c) are included to support the new Stack structure utilized in the fix.

The fix was tested using the repro provided in the NIST report ([repro also linked here](https://github.com/htacg/tidy-html5/issues/946)). Below is the output from the repro before and after the fix is applied. As well as a valgrind evaluation of both versions.

repro, pre-fix, via `tidy -g test.html`:
https://pastebin.com/xppXfDv8

Valgrind of repro, pre-fix, via `valgrind tidy -g test.html`:
```
==4399==
==4399== HEAP SUMMARY:
==4399==     in use at exit: 0 bytes in 0 blocks
==4399==   total heap usage: 2,383 allocs, 2,391 frees, 2,510,491 bytes allocated
==4399==
==4399== All heap blocks were freed -- no leaks are possible
==4399==
==4399== For lists of detected and suppressed errors, rerun with: -s
==4399== ERROR SUMMARY: 447 errors from 40 contexts (suppressed: 0 from 0)
```
repro, post-fix, via `tidy -g test.html`:
https://pastebin.com/DSBV9U4X

Valgrind of repro, post fix, via `valgrind tidy -g test.html`:
```
==1954==
==1954== HEAP SUMMARY:
==1954==     in use at exit: 0 bytes in 0 blocks
==1954==   total heap usage: 3,200 allocs, 3,200 frees, 3,788,259 bytes allocated
==1954==
==1954== All heap blocks were freed -- no leaks are possible
==1954==
==1954== For lists of detected and suppressed errors, rerun with: -s
==1954== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

###### Change Log  

- CVE-2021-33391

###### Does this affect the toolchain?
  

**NO**

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-2021-33391

###### Test Methodology

- [Pipeline build](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=353241&view=artifacts&pathAsName=false&type=publishedArtifacts)

### CODE REVIEW

File: SPECS/tidy/tidy.spec [hunk 1]
```diff
@@ -11,6 +11,7 @@ Vendor:         Microsoft Corporation
 Distribution:   Mariner
 URL:            https://www.html-tidy.org/
 Source0:        https://github.com/htacg/%{upname}/archive/%{version}.tar.gz#/%{upname}-%{version}.tar.gz
+patch0: CVE-2021-33391.patch
```
Comment (christopherco) [inline SPECS/tidy/tidy.spec:14]:
```suggestion
Patch0:         CVE-2021-33391.patch
```

File: SPECS/tidy/tidy.spec
Comment (christopherco) [inline SPECS/tidy/tidy.spec:1]: Bump release version & add changelog entry

**Review by christopherco — CHANGES_REQUESTED** [review]

File: SPECS/tidy/tidy.spec [hunk 2]
```diff
@@ -11,6 +11,7 @@ Vendor:         Microsoft Corporation
 Distribution:   Mariner
```
Comment (mfrw) [inline SPECS/tidy/tidy.spec:11]: Let's please change the PR title to be more changelog friendly. :)

File: SPECS/tidy/tidy.spec [hunk 3]
```diff
@@ -5,12 +5,13 @@
...
 Name:           tidy
 Version:        5.8.0
-Release:        5%{?dist}
+Release:        6%{?dist}
 License:        W3C
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
 URL:            https://www.html-tidy.org/
 Source0:        https://github.com/htacg/%{upname}/archive/%{version}.tar.gz#/%{upname}-%{version}.tar.gz
+patch0:         CVE-2021-33391.patch
```
Comment (christopherco) [inline SPECS/tidy/tidy.spec:14]:
```suggestion
Patch0:         CVE-2021-33391.patch
```
nit

File: SPECS/tidy/tidy.spec [hunk 4]
```diff
@@ -92,6 +93,9 @@ rm -fv %{buildroot}%{_libdir}/libtidy.a
 %{_libdir}/pkgconfig/tidy.pc
 
 %changelog
+* Mon May 01 2023 Sean Dougherty <sdougherty@microsoft.com> - 5.8.0-6
+- Added bespoke backport patch to fix CVE-2021-33391
```
Comment (christopherco) [inline SPECS/tidy/tidy.spec:97]:
```suggestion
- Backported patch to fix CVE-2021-33391
```
nit - suggestion for more concise wording

**Review by christopherco — APPROVED** [review]

**Pull request merged**